### PR TITLE
Updated URL to scrape opinions for New Hampshire Supreme court

### LINKED ISSUE #444: Fix opinion scraper for New Hampshire Supreme court
The web site for the **New Hampshire Supreme Court** changed.
The URL structure to seach opinions is different, as well as the list structure with the opinions.
https://www.courts.nh.gov/our-courts/supreme-court/orders-and-opinions/opinions

`juriscraper.opinions.united_states.state.nh`

### PR DESCRIPTION
**Pull request by satsuki-chan**
Updated scraper to read information directly from the API for the new web site for the **New Hampshire Supreme Court**.
https://www.courts.nh.gov/content/api/documents

Fixes #444

### CODE REVIEW

File: juriscraper/opinions/united_states/state/nh.py [hunk 1]
```diff
@@ -1,127 +1,58 @@
...
-                        # docket and name (root) should be contained in first substring
-                        error = f"Invalid anchor root string format: {text}"
-                        raise InsanityException(error)
-                    # no docket in the substring, its a trailing name substring
-                    # that they broke over multiple lines, so glue it to the
-                    # previous name substring
-                    name_substrings[i - 1] += f" {text}"
-            pairs.append(
+        self.url = "https://www.courts.nh.gov/content/api/documents"
+        self.method = "GET"
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:30]: Is this the default?  Can you verify is this is necessary

File: juriscraper/opinions/united_states/state/nh.py [hunk 2]
```diff
@@ -1,127 +1,58 @@
...
+        self.parameters = {
+            "sort": "field_date_posted|desc",
+            "page": "1",
+            "size": "25",
+            "purpose": "1331+undefined",
+            "tag": "1206+",
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:36]: Also - a comment explaining these values may be helpful.  

I think tag is important but im not sure purpose is.

File: juriscraper/opinions/united_states/state/nh.py [hunk 3]
```diff
@@ -1,127 +1,58 @@
...
+        }
+
+    def _download(self, request_dict={}):
+        if self.test_mode_enabled():
+            return super()._download(request_dict)
+        resp = self.request["session"].get(self.url, params=self.parameters)
+        return resp.json()
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:43]: @mlissner --- can I get a second opinion on this.  Is this a fine way to use _download.  I suggested it to @satsuki-chan earlier but I would like your thoughts.

**Review by flooie — CHANGES_REQUESTED** [review]
@satsuki-chan this looks pretty nice and so much cleaner.  

Can you just double check a few things.

File: juriscraper/opinions/united_states/state/nh.py (hunk 1 above)
Comment (flooie) [reply juriscraper/opinions/united_states/state/nh.py:30]: The GET method I mean.

File: juriscraper/opinions/united_states/state/nh.py (hunk 3 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:43]: I have the same question @mlissner  👆
It was the only way I found to load the example file to run the test for this scraper.

File: juriscraper/opinions/united_states/state/nh.py (hunk 2 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:36]: `"purpose"` must be included because it filter out other types of documents that aren't opinions. Opinions have a `"purpose"` value of **1331|**, while other documents have values: **1856|**, **1491|** and **1691|**
For example:
`
{'documentContent': "<div class='document__detail'>\n\t\t\t\t<div class='document__detail__icon__container'>\n\t\t\t\t\t\n\t\t\t\t</div>\n\t\t\t\t<div class='document__detail__container'>\n\t\t\t\t\t<div class='document__detail__title'><h3><a href='https://www.courts.nh.gov/sites/g/files/ehbemt471/files/documents/2021-11/20200490.pdf' >2020-0490, Mary Feeney v. Karyn Kelley</a></h3></div>\n\t\t\t\t\t<div class='document__detail__description'></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Document Format: </strong>PDF</div></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Date: </strong>10/20/2021</div></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t</div>\n\t\t\t</div>", 'documentNumber': None, 'documentJudge': None, 'documentFiled': None, 'documentPosted': '2021-10-20', 'documentRevised': None, 'documentParties': None, 'documentCategories': None, 'documentPurposes': **'1856|'**, 'documentTags': None, 'documentName': '2020-0490, Mary Feeney v. Karyn Kelley'}`

`
{'documentContent': "<div class='document__detail'>\n\t\t\t\t<div class='document__detail__icon__container'>\n\t\t\t\t\t\n\t\t\t\t</div>\n\t\t\t\t<div class='document__detail__container'>\n\t\t\t\t\t<div class='document__detail__title'><h3><a href='https://www.courts.nh.gov/sites/g/files/ehbemt471/files/documents/2021-10/12-10-21-public-hearing-notice-and-order.pdf' >Public hearing notice and order seeking comment December 10, 2021 public hearing</a></h3></div>\n\t\t\t\t\t<div class='document__detail__description'></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Document Format: </strong>PDF</div></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Date: </strong>10/04/2021</div></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t</div>\n\t\t\t</div>", 'documentNumber': None, 'documentJudge': None, 'documentFiled': None, 'documentPosted': '2021-10-04', 'documentRevised': None, 'documentParties': None, 'documentCategories': None, 'documentPurposes': **'1491|'**, 'documentTags': None, 'documentName': 'Public hearing notice and order seeking comment December 10, 2021 public hearing'}`

`
{'documentContent': "<div class='document__detail'>\n\t\t\t\t<div class='document__detail__icon__container'>\n\t\t\t\t\t\n\t\t\t\t</div>\n\t\t\t\t<div class='document__detail__container'>\n\t\t\t\t\t<div class='document__detail__title'><h3><a href='https://www.courts.nh.gov/sites/g/files/ehbemt471/files/documents/2021-09/superior-court-administrative-order-nubmer-2021-006.pdf' >Administrative Order 2021 - 06: Obsolete Administrative Orders</a></h3></div>\n\t\t\t\t\t<div class='document__detail__description'></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Document Format: </strong>PDF</div></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Date: </strong>09/30/2021</div></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t</div>\n\t\t\t</div>", 'documentNumber': None, 'documentJudge': None, 'documentFiled': None, 'documentPosted': '2021-09-30', 'documentRevised': None, 'documentParties': None, 'documentCategories': None, 'documentPurposes': **'1691|'**, 'documentTags': None, 'documentName': 'Administrative Order 2021 - 06: Obsolete Administrative Orders'}
`

I'll add a comment about this in the parameters for the call to the API.

Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:36]: So far, I have't found an opinion file withou a `"tag"` value of **1206|**, but this parameter can also have the value **None**. I recommend to keep it.

`{'documentContent': "<div class='document__detail'>\n\t\t\t\t<div class='document__detail__icon__container'>\n\t\t\t\t\t\n\t\t\t\t</div>\n\t\t\t\t<div class='document__detail__container'>\n\t\t\t\t\t<div class='document__detail__title'><h3><a href='https://www.courts.nh.gov/sites/g/files/ehbemt471/files/documents/2021-11/20210008.pdf' >2021-0008, In the Matter of Jennifer Juri and  Michael Juri</a></h3></div>\n\t\t\t\t\t<div class='document__detail__description'></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Document Format: </strong>PDF</div></div>\n\t\t\t\t\t<div class='document__detail__information'><div><strong>Date: </strong>10/12/2021</div></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t\t<div class='document__detail__information'></div>\n\t\t\t\t</div>\n\t\t\t</div>", 'documentNumber': None, 'documentJudge': None, 'documentFiled': None, 'documentPosted': '2021-10-12', 'documentRevised': None, 'documentParties': None, 'documentCategories': None, 'documentPurposes': '1856|', 'documentTags': None, 'documentName': '2021-0008, In the Matter of Jennifer Juri and  Michael Juri'}
`

I'll also add a comment about param value.

File: juriscraper/opinions/united_states/state/nh.py [hunk 4]
```diff
@@ -1,127 +1,58 @@
...
+            "page": "1",
+            "size": "25",
+            "purpose": "1331+undefined",
+            "tag": "1206+",
+        }
+
+    def _download(self, request_dict={}):
+        if self.test_mode_enabled():
+            return super()._download(request_dict)
+        resp = self.request["session"].get(self.url, params=self.parameters)
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:42]: Additionally, some comments explaining what is happening here probably make sense.  Like 
This site pings the API instead of the homepage to get the JSON data..

File: juriscraper/opinions/united_states/state/nh.py (hunk 1 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Removed

File: juriscraper/opinions/united_states/state/nh.py (hunk 4 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:44]: Done

Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:44]: Done

File: juriscraper/opinions/united_states/state/nh.py [hunk 5]
```diff
@@ -1,127 +1,62 @@
...
+            return super()._download(request_dict)
+        # Calls to API with POST are not allowed, using GET method
+        resp = self.request["session"].get(self.url, params=self.parameters)
+        return resp.json()
+
+    def _process_html(self) -> None:
+        for case in self.html["data"]:
+            docket, name = case["documentName"].split(",", 1)
+            date = case["documentPosted"]
+            url = re.findall(r"(http.*docu.*\.pdf)", case["documentContent"])[
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:51]: Can you do some more testing for the URL and docket, name generation.  

I was able to break the scraper by paging back only one page.   I think the split on , is too fragile.
Also - There is a case, `Hampstead School Board & a. v. School Administrative Unit No. 55` that you should look into.  

If you check out the url you can see its not an opinion its an order withdrawing the previous opinion from April.  The scraper assigns the new order the same values as the previous opinion.  And it extracts the url from the "related_document section" of the json response.  

This is close but it needs a little more testing and review.

File: juriscraper/opinions/united_states/state/nh.py [hunk 6]
```diff
@@ -1,127 +1,62 @@
...
-            pairs.append(
+        self.url = "https://www.courts.nh.gov/content/api/documents"
+        self.parameters = {
+            "sort": "field_date_posted|desc",
+            "page": "1",
+            "size": "25",
+            # Opinions have a "documentPurposes" value of "1331|"
+            "purpose": "1331+undefined",
+            # Opinions have a "documentTags" value of "1206|"
+            "tag": "1206+",
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:37]: Thanks for adding these, but I wonder if these comments are going to make much sense in the future when we are re-reviewing this code.  And what do the + or +undefined really mean.  Do they mean these values and anything else?

**Review by flooie — CHANGES_REQUESTED** [review]
Thanks for your changes, I think we have a few things to fix on this one.

File: juriscraper/opinions/united_states/state/nh.py [hunk 7]
```diff
@@ -1,127 +1,62 @@
...
+        }
+
+    def _download(self, request_dict={}):
+        if self.test_mode_enabled():
+            return super()._download(request_dict)
+        # Calls to API with POST are not allowed, using GET method
```
[the remark this answers is not in the document]
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:43]: Only request with the **POST** method include the values defined in `self.parameters` in the requests:

https://github.com/freelawproject/juriscraper/blob/3ad379e9b7ff658fd19172fdba510aab217a31f2/juriscraper/AbstractSite.py#L359-L369

Requests with **GET** ignore them:
https://github.com/freelawproject/juriscraper/blob/3ad379e9b7ff658fd19172fdba510aab217a31f2/juriscraper/AbstractSite.py#L346-L357

File: juriscraper/opinions/united_states/state/nh.py (hunk 3 above)
Comment (mlissner) [reply juriscraper/opinions/united_states/state/nh.py]: TL;DR: Let's make ourselves happy the next time we are dealing with JSON.

------

I think the big picture here is that we don't have a good way to deal with JSON? If that assumption is correct, yeah, the above would *work*, but I think this is a common-enough use case that instead of working around our own code with hacks like this, we should add functionality to handle this in an elegant way. 

I think in that case, that could mean a change to the `_download` function itself or maybe something else in the AbstractSite. I haven't looked closely enough to know the right solution.

It pains me to see this function abused like this to work around our own design assumptions (that there'd never be JSON coming from courts). Instead of hacking around our own code base, @satsuki-chan would you mind proposing a solution that's actually good?

Comment (flooie) [reply juriscraper/opinions/united_states/state/nh.py]: Thanks @mlissner

Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: @mlissner , the current architecture _**is able to process JSON responses**_ in method `_return_response_text_object`:
https://github.com/freelawproject/juriscraper/blob/23e61f8a391c9d3c68b9194168c9ee10e6a7ae85/juriscraper/AbstractSite.py#L382-L388

What is missing is that the optional parameters for page requests in dictionary `self.parameters` are _**only included when the requests is made with POST**_ method:
https://github.com/freelawproject/juriscraper/blob/23e61f8a391c9d3c68b9194168c9ee10e6a7ae85/juriscraper/AbstractSite.py#L359-L369

To solve this kind of issues, is necessary to also add the parameter `data` in the **GET** request in method `_request_url_get`:
https://github.com/freelawproject/juriscraper/blob/23e61f8a391c9d3c68b9194168c9ee10e6a7ae85/juriscraper/AbstractSite.py#L346-L357

```
def _request_url_get(self, url):
    """Execute GET request and assign appropriate request dictionary
    values
    """
    self.request["url"] = url
    self.request["response"] = self.request["session"].get(
        url,
        headers=self.request["headers"],
        verify=self.request["verify"],
        data=self.parameters,
        timeout=60,
        **self.request["parameters"],
    )
```

Given the scope and efect that his change might have in all other scrapers, I propose to make this change in a separate PR with fixes for all scrapers that report issues with this change in their GET requests.

Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: To clarify, POST requests to the New Hampshire API are not allowed. In my tests with the POST methods I got the HTTP code **405**, Method not allowed.

Comment (mlissner) [reply juriscraper/opinions/united_states/state/nh.py]: I see. If that's all that's needed, I'm fine with adding that in a separate PR, but in the meantime, why not just add the GET parameters to the URL?

File: juriscraper/opinions/united_states/state/nh.py (hunk 6 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Added more detailed comments and examples.
Could you review them @flooie ?

File: juriscraper/opinions/united_states/state/nh.py (hunk 5 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Added validation to only extract files for opinions.

File: juriscraper/opinions/united_states/state/nh.py (hunk 3 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Done.

File: juriscraper/opinions/united_states/state/nh.py [hunk 8]
```diff
@@ -1,127 +1,142 @@
...
+        """
+        name_re = (
+            r"(?P<docket>(((LD-|)\d{2,4}-\d{3,},\s*|)(LD-|)\d{2,4}-\d{3,}(,\s*(&|and)|\s*(&|and)|,)\s*|)(LD-|)\d{2,4}-\d{3,})"
+            r"(?P<name>.*)"
+        )
+        name_sec_re = (
+            r"(?P<name_1>.*)"
+            r"(?P<docket_2>(\s*(&|and)|,)\s*(LD-|)\d{2,4}-\d{3,})"
+            r"(?P<name_2>.*)"
         )
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:87]: I see what you are doing but this is overly complicated.

File: juriscraper/opinions/united_states/state/nh.py [hunk 9]
```diff
@@ -1,127 +1,142 @@
...
+                        .strip()
+                        .lstrip(",")
+                        .strip()
+                    )
+                    name_sec = (
+                        docket_name_sec.group("name_2")
+                        .strip()
+                        .lstrip(",.")
+                        .strip()
+                    )
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:119]: same here... im reading this and I don't understand what docket_name_sec is

File: juriscraper/opinions/united_states/state/nh.py [hunk 10]
```diff
@@ -1,127 +1,142 @@
...
+            * 2018: subcategory=1601    * 2017: subcategory=1596
+            * 2016: subcategory=1591    * 2015: subcategory=1586
+            * 2014: subcategory=1581    * 2013: subcategory=1576
+            * 2012: subcategory=1571    * 2011: subcategory=1566
+            * 2010: subcategory=1561    * 2009: subcategory=1556
+        Only one can have a key number; using both with a value in the same
+        request (even with a valid key) returns zero documents.
+        Using none, returns all documents found with other parameters values.
+        Examples:
+            - https://www.courts.nh.gov/content/api/documents?text=&category=+undefined&purpose=1331+undefined&tag=+&subcategory=1556&sort=field_date_posted|desc&page=3&size=10
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:45]: So does this mean everything will break in 2022 if we use these but we can use them to create a back scraper?

File: juriscraper/opinions/united_states/state/nh.py (hunk 8 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Moved to a function for simplification.

File: juriscraper/opinions/united_states/state/nh.py (hunk 9 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Moved to a function for simplification

File: juriscraper/opinions/united_states/state/nh.py (hunk 10 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: The `tag` and  `subcategory` parameters are only necesary when searching documents for an specific year. So far, those are the only years with an assigned key number that I could finds the court's web site.
Comments edited for clarification.

File: juriscraper/opinions/united_states/state/nh.py [hunk 11]
```diff
@@ -1,127 +1,150 @@
...
-        number and a single case name.  However, there are two other formats
-        we've seen and must work around.
+    def name_docket(self, title) -> Tuple[str, str]:
+        """
+        Expected title values:
+            - 2020-0544 and 2020-0554, In re G.B.
+            - 2020-0528, 2020-0546, 2020-0548, In re S.A. & a.
+            - 2019-0283 In re R.M.
+            - 2017-0336 & 2019-0071  State of New Hampshire v. Daniel Turcotte
+            - 2014-0315, Deere & Company & a. v. The State of New Hampshire 2014-0441, Kubota Tractor Corporation v. The State of New Hampshire 2014-0575, Husqvarna Professional Products, Inc. v. The State of New Hampshire
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:90]: This method seems unnecessary.  And overly complicated.

File: juriscraper/opinions/united_states/state/nh.py [hunk 12]
```diff
@@ -1,127 +1,150 @@
...
+            - 2013-0779, Ichiban Japanese Steakhouse, Inc. v. Kymberly Rocheleau, and 2013-0780, Ichiban Japanese Steakhouse, Inc. v. Samantha Greaney
+            - 2013-0513, Greg DuPont v. Nashua Police Department (consolidated with 2014-0017, Gregory DuPont v. Peter McDonough & a.)
+            - 2013-0403, 2013-0445, and 2013-0593, In re Guardianship of Madelyn B.
+            - 2013-0392. State of New Hampshire v. Kevin Balch
+            - 010-290  Eric Lee Knight v. Cheryl Ann Maher
+            - 2010-233/2010-234 Appeal of the Hartford Insurance Company (New Hampshire Compensation Appeals Board)
+            - 10-436, New Hampshire Health Care Association & a. v. Governor & a.
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:97]: I appreciate the effort you put into collecting these edge cases - but I think none of this should be here - if we were to even keep this function.  This feels like a testing method you would use to test the parser.

File: juriscraper/opinions/united_states/state/nh.py [hunk 13]
```diff
@@ -1,127 +1,150 @@
...
-    def _get_docket_numbers(self):
-        return [pair["docket"] for pair in self.docket_name_pairs]
+            date = case["documentPosted"]
+            if not date:
+                logger.info(f"Opinion without date: {case}")
+                continue
 
-    def _get_anchor_docket_name_pairs(self):
-        """The court has some ugly HTML practices that we need to handle.
+            name, docket = self.name_docket(case["documentName"])
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:73]: You should just replace this line and method with the following.

```suggestion
            # In juvenile cases - the case name contains the docket number
            # So we use a negative lookbehind to ignore docket numbers after
            # the word juvenile.
            dockets = re.findall(
                r"(?<!JUVENILE )((\w+-)?\d{4}-\d+)", case["documentName"]
            )
            docket = ", ".join([docket[0] for docket in dockets])

            # Strip out the docket numbers and clean up the case names.
            name = re.sub(
                r"(?<!JUVENILE )((\w+-)?\d{4}-\d+)", "", case["documentName"]
            )
            name = name.strip(", ").replace(", and ,", " and ")

```

File: juriscraper/opinions/united_states/state/nh.py (hunk 11 above)
Comment (flooie) [reply juriscraper/opinions/united_states/state/nh.py:90]: I would suggest deleting it and using my suggested code.

File: juriscraper/opinions/united_states/state/nh.py [hunk 14]
```diff
@@ -1,127 +1,150 @@
...
+    found with the other parameters values.
+    For this scraper, filtering documents per year is not necesary.
+    Examples:
+        - https://www.courts.nh.gov/content/api/documents?text=&category=+undefined&purpose=1331+undefined&tag=+&subcategory=1556&sort=field_date_posted|desc&page=3&size=10
+        - https://www.courts.nh.gov/content/api/documents?text=&category=+undefined&purpose=1331+undefined&tag=1366+&subcategory=&sort=field_date_posted|desc&page=1&size=25
+"""
 
 import re
-from datetime import date
+from typing import Tuple
```
Comment (flooie) [inline juriscraper/opinions/united_states/state/nh.py:40]: If you remove that method you can delete this import as well.

**Review by flooie — CHANGES_REQUESTED** [review]
@satsuki-chan back to you.

File: juriscraper/opinions/united_states/state/nh.py (hunk 11 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Added your suggested code to the function.

File: juriscraper/opinions/united_states/state/nh.py (hunk 12 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py]: Removed the comments and added tests examples with edge cases.

File: juriscraper/opinions/united_states/state/nh.py (hunk 13 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:73]: Used code in new version for method `self.name_docket` to handle some edge cases.

File: juriscraper/opinions/united_states/state/nh.py (hunk 14 above)
Comment (satsuki-chan) [reply juriscraper/opinions/united_states/state/nh.py:40]: Updated the code for the method, so the import remains.
